add -Wconf flag for configurable warnings, @nowarn annotation for local suppression

lrytz opened the pull request:
This PR adds a `-Wconf` compiler flag that allows filtering and configuring compiler warnings (silence them, or turn them into errors).

It also integrates the fantastic [silencer](https://github.com/ghik/silencer) compiler plugin by @ghik into the compiler, which allows suppressing warnings locally using the `@nowarn` annotation.

### Configurable Warnings

Warnings are assigned a category.

Warnings can be filtered by category, by message regex, by site where they are issued, and by source path. Deprecations can additionally be filtered by origin (deprecated definition) and `since` version.

Filtered warnings can be reported as error, warning, info, summary (like deprecations) or silent.

### Local Suppression

The `@nowarn` annotation suppresses warnings within the scope covered by the annotation.
  - `@nowarn def foo = ...`, `@nowarn class C { ... }`: suppress warnings in a definition
  - `expression: @nowarn`: suppress warnings in a specific expression

The annotation can be configured to filter selected warnings, for example `@nowarn("cat=deprecation")` only suppresses deprecation warnings. The filter configuration syntax is the same as in `-Wconf`.

### Help Text

```
$> scalac -Wconf:help
Configure compiler warnings.
Syntax: -Wconf:<filters>:<action>,<filters>:<action>,...
multiple <filters> are combined with &, i.e., <filter>&...&<filter>

Note: Run with `-Wconf:any:warning-verbose` to print warnings with their category, site,
and (for deprecations) origin and since-version.

<filter>
  - Any message: any

  - Message categories: cat=deprecation, cat=lint, cat=lint-infer-any
    The full list of warning categories is shown at the end of this help text.

  - Message content: msg=regex
    The regex need only match some part of the message, not all of it.

  - Site where the warning is triggered: site=my\.package\..*
    The regex must match the full name (`package.Class.method`) of the warning position.

  - Source file name: src=src_managed/.*
    If `-rootdir` is specified, the regex must match the canonical path relative to the
    root directory. Otherwise, the regex must match the canonical path relative to any
    path segment (`b/.*Test.scala` matches `/a/b/XTest.scala` but not `/ab/Test.scala`).
    Use unix-style paths, separated by `/`.

  - Origin of deprecation: origin=external\.package\..*
    The regex must match the full name (`package.Class.method`) of the deprecated entity.

  - Since of deprecation: since<1.24
    Valid operators: <, =, >, valid versions: N, N.M, N.M.P. Compares against the first
    version of the form N, N.M or N.M.P found in the `since` parameter of the deprecation,
    for example `1.2.3` in `@deprecated("", "some lib 1.2.3-foo")`.

<action>
  - error / e
  - warning / w
  - warning-summary / ws (summary with the number of warnings, like for deprecations)
  - warning-verbose / wv (show warning category and site)
  - info / i             (infos are not counted as warnings and don't affect `-Werror`)
  - info-summary / is
  - info-verbose / iv
  - silent / s

The default configuration is:
  -Wconf:cat=deprecation:ws,cat=feature:ws,cat=optimizer:ws

User-defined configurations are added to the left. The leftmost rule matching
a warning message defines the action.

Examples:
  - change every warning into an error: -Wconf:any:error
  - silence certain deprecations: -Wconf:origin=some\.lib\..*&since>2.2:s

Full list of message categories:
 - deprecation
 - feature, feature-dynamics, feature-existentials, feature-higher-kinds, feature-implicit-conversions, feature-macros, feature-postfix-ops, feature-reflective-calls
 - java-source
 - lint, lint-adapted-args, lint-byname-implicit, lint-constant, lint-delayedinit-select, lint-deprecation, lint-doc-detached, lint-eta-sam, lint-eta-zero, lint-implicit-not-found, lint-inaccessible, lint-infer-any, lint-missing-interpolator, lint-nonlocal-return, lint-nullary-override, lint-nullary-unit, lint-option-implicit, lint-package-object-classes, lint-poly-implicit-overload, lint-private-shadow, lint-recurse-with-default, lint-serial, lint-stars-align, lint-type-parameter-shadow, lint-unit-specialization
 - optimizer
 - other, other-debug, other-match-analysis, other-migration, other-pure-statement, other-shadowing
 - scaladoc
 - unchecked
 - unused, unused-imports, unused-locals, unused-nowarn, unused-params, unused-pat-vars, unused-privates
 - w-flag, w-flag-dead-code, w-flag-extra-implicit, w-flag-numeric-widen, w-flag-self-implicit, w-flag-value-discard

To suppress warnings locally, use the `scala.annotation.nowarn` annotation.

Note: on the command-line you might need to quote configurations containing `*` or `&`
to prevent the shell from expanding patterns.
```

## Significant changes

  - What's described above in the PR description
  - MiMa exception for addition of `scala.annotation.nowarn`
  - Adds a `-rootdir` compiler flag. It is used to relativize file paths when using source filters (`-Wconf:src=some/source/File.scala:s`), there might be other uses for it in the future.
  - Compiler API changes
    - `warning` methods in currentRun.reporting and context reporter take new `Category` parameter. The `global.reporter` is unchanged, warnings reported through it are not filtered by `Wconf`.
    - Removed a few deprecated methods (compilationUnit.error/warning)
  - Reporting warnings is now delayed until after typer, because the typer collects `@nowarn` annotations.
  - If we stop before typer (e.g., because there are errors in the parser), all warnings are shown, even if they are enclosed in `@nowarn`
  - If a phase before typer has both errors and warnings, all errors are printed before the warnings.
  - Unchecked warnings are now all shown by default (they were summarized like deprecations before)
  - The `-deprecation`, `-feature` and `-unchecked` settings are no longer directly used in the compiler, they are shortcuts for specific `Wconf` configurations. the compiler only looks at `-Wconf`.

lrytz commented:
Message filtering is performent in `global.currentRun.reporting`, not in the `Global.reporter`.

The reasons are:
  - Separation of concerns: `Reporter`s worry about how to do reporting, removing duplicate messages.
  - Custom `Reporter`s are used by sbt, silencer, REPL, etc. It's too hard to do the necessary changes to the `Reporter` interface.
  - The `Wconf` setting could change between compiler runs. `currentRun.reporting` respects those changes.

So all warnings in the compiler should go through `global.runReporting` (which is the same as `global.currentRun.reporting`). This method takes four parameters: pos, msg, category (new), site (new). The site is usually `context.owner` (in the frontend) or `currentOwner` (in transformations).

`Context` has a `warn` method with 3 parameters (pos, msg, category) and inserts the `owner` as `site`, so this is used in the frontend (context reporters are also used to support silent mode / trying twice).

The `global.warning` method is deprecated and no longer used.

There are a few calls to `Reporter.warning` left in the codebase where no `runReporting` is reachable, I think they are aoll OK not to categorize. E.g., when running Scaladoc

```
    else if (docSettings.showPlugins.value)
      reporter.warning(null, "Plugins are not available when using Scaladoc")
```

lrytz commented:
Added support for local warning suppression using `@silent`. I borrowed heavily from silencer, in design, implementation and tests; cc @ghik, a huge THANK YOU for your work!

ghik commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1546,6 +1546,9 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
           printStatisticsFor(phase)
 
+        if (!globalPhase.hasNext)
+          runReporting.warnUnusedSuppressions()
```
Is this done unconditionally? In silencer, there was a flag to enable it. I think doing it unconditionally may cause problems in cross compiled code (against multiple scala versions or backends) where a warning is not emitted for all compilation targets.

lrytz replied:
That's a good point. But note that you can filter these `unused @silent` warnings with a `-Wconf` setting (I'll add a `UnusedSilent` category, instead of just using `Other`), so I think it should be fine to report them unconditionally. Note that they are warnings, not errors.

lrytz replied:
I added `-Wunused:silent` in the end

ijuma commented:
This will be immensely useful. Thanks for working on this!

ghik commented on src/library/scala/annotation/silent.scala:
```diff
@@ -0,0 +1,32 @@
+  *   def f = { 1; deprecated() } // show deprecation warning
+  * }}}
+  */
+class silent(value: String = "") extends ConstantAnnotation
```
I wonder if we can avoid writing the warning configuration as a raw string.
For example, I would prefer something like `@silent(msg = "deprecated")` instead of `@silent("msg=deprecated")`.
I know that one of the hurdles here is that named/default arguments don't work very well with annotations, i.e.

```
Welcome to Scala 2.12.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_201).
Type in expressions for evaluation. Or try :help.

scala> class annot(first: String = "", second: String = "") extends scala.annotation.Annotation
defined class annot

scala> @annot(second = "second") class Annotated
<console>:13: warning: Usage of named or default arguments transformed this annotation
constructor call into a block. The corresponding AnnotationInfo
will contain references to local values and default getters instead
of the actual argument trees
       @annot(second = "second") class Annotated
        ^
defined class Annotated
```

But maybe this is a good opportunity to fix that?

lrytz replied:
Right, I just filed https://github.com/scala/bug/issues/11724. On the other hand,
  - the syntax now is exactly the same as for `-Wconf`
  - the string supports conjunctions (`cat=deprecation&origin=my.library.*`)

lrytz commented:
Removed the WIP flag and update the PR description, this is ready to be reviewed!

eed3si9n commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -24,6 +26,72 @@ trait Warnings {
+    "Configure reporting of compiler warnings; use `help` for details.",
+    helpText = Some(
+      s"""Configure compiler warnings.
+         |Syntax: -Wconf:<filter>&...&<filter>:<action>,<filter>:<action>,...
```
```suggestion
         |Syntax: -Wconf:<filter>:<action>,<filter>:<action>,...
```

The `&` threw me off since I wasn't sure about the precedence of `&`, `:`, and `,`. Maybe `&` that can be added as a section below as "`<filter>` might be chained etc.".

lrytz replied:
Thanks, that looks better. I'm also happy to change `&` for some other character, but this is the best I could come up with. 

The first `:` (in `-Wconf:`) and the `,` to split settings is given by the MultiStringSetting. There's also the "non-colonated-space-separated mode" that eats until the next setting (`-Wconf <filters>:<action> <filters>:<action> -nextSetting`), but it's not recommended.

The `:` before `<action>` and `&` to conjunct filters could be changed.

ghik commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -447,10 +453,13 @@ object Reporting {
+      private[this] val cache = mutable.Map.empty[SourceFile, Boolean]
+
+      def matches(message: Message): Boolean = cache.getOrElseUpdate(message.pos.source, {
+        var sourcePath = message.pos.source.file.canonicalPath.replace("\\", "/")
```
Should we relativize paths against some root directories? Without that, builds technically will not be reproducible because absolute paths can be different on every system. I suppose there is a non-zero chance that this may cause problems - for example, a regex may accidentally match a name of some parent directory well outside of the project structure.

lrytz replied:
I guess you're right... one root directory should be enough, no? What do you think about keeping the current behavior if no root directory is specified?

ghik replied:
That's exactly what silencer does.

ghik replied:
I wonder if the concept of root source directory could be useful elsewhere in the compiler than just for silencing warnings.

sjrd replied:
It would definitely be useful to Scala.js, to store source positions that are not absolute.

lrytz replied:
Does Scala.js not have any custom compiler flags?

I guess the most sensible root directory would be the project root, i.e., the path of the git checkout. `-rootdir`.

For `-Wconf`, we could also interpret `src=/...` as a full absolute regex if it starts with `/`, but that would mean repeating the project root path in every `src` filter.

sjrd replied:
> Does Scala.js not have any custom compiler flags?

Yes, we do, but they're only useful if people use them. ;)

lrytz replied:
Added the `-rootdir` setting in the latest commit.

ghik commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -35,16 +35,35 @@ trait Warnings {
+         |    The regex must match the full name of the warning position.
+         |
+         |  - Source file name: src=src_managed/.*
+         |    The regex must match the canonical path, starting at any path segment.
```
The documentation should definitely mention that file paths will always have file name separators normalized to Unix-style (`/`).

lrytz commented:
@SethTisue would you be interested to review this, at least on a high level? I updated the PR description, and also included a section on significant changes.

ghik commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -54,8 +54,10 @@ trait Warnings {
          |  - Source file name: src=src_managed/.*
-         |    The regex must match the canonical path, starting at any path segment.
-         |    Example: `b/.*Test.scala` matches `/a/b/XTest.scala` but not `/ab/YTest.scala`.
+         |    If `-rootdir` is specified, the regex must match the canonical path relative to the
```
We should also mention that file name separators are normalized to `/`

lrytz replied:
3 lines below it says "Use unix-style paths, separated by `/`" - is that it, did you mean something else?

ghik replied:
Ugh, sorry, no idea how I missed that.

ghik commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -544,16 +546,21 @@ object Reporting {
-        regex(pat).map(SourcePattern)
+        val arg = s.substring(4)
+        var pat = new mutable.StringBuilder()
+        if (rootDir.nonEmpty) pat += '^' ++= rootDir
```
Would it not be easier to first relativize the path, then normalize filename separators to `/` and finally match against the pattern instead of modifying the regex itself?

lrytz replied:
Either way seems fine with me, so I'll keep this as-is. We're building a new regex anyways.

ghik commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -544,16 +546,21 @@ object Reporting {
-        regex(pat).map(SourcePattern)
+        val arg = s.substring(4)
+        var pat = new mutable.StringBuilder()
+        if (rootDir.nonEmpty) pat += '^' ++= rootDir
```
`rootDir` needs to be escaped with `Regex.quote`

lrytz replied:
Thanks, done.

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -14,102 +14,212 @@ package scala
 /** Provides delegates to the reporter doing the actual work.
  * PerRunReporting implements per-Run stateful info tracking and reporting
- *
- * TODO: make reporting configurable
```
:rocket:

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -14,102 +14,212 @@ package scala
-            |See the Scaladoc for value $fqname for a discussion
-            |why the feature $req be explicitly enabled.""".stripMargin
-      )
+          s"""
```
Maybe it's easier to achieve the desired indentation alignment with `sm"""` because of the extra letter. Is it appealing with vbar in even column, or to left of even column?

lrytz replied:
I don't have a personal opinion here, my only opinion is "do what ever IntelliJ does", so I don't get any spurious changes when editing this code (in IntelliJ)...

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -142,3 +270,334 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
+    object LintSerial extends Lint; add(LintSerial)
+    object LintEtaZero extends Lint; add(LintEtaZero)
+    object LintEtaSam extends Lint; add(LintEtaSam)
+    object LintDeprecation extends Lint; add(LintDeprecation)
```
Is there an easy mechanism to reduce redundancy? Easy in the sense of impossible is doable and difficult is easy.

lrytz replied:
This is the best I could come up with.. We cannot put `add(this)` it in the constructor of `WarningCategory`, as the objects here are evaluated lazily.

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -142,3 +270,334 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
+    object WFlagDeadCode extends WFlag; add(WFlagDeadCode)
+    object WFlagExtraImplicit extends WFlag; add(WFlagExtraImplicit)
+    object WFlagNumericWiden extends WFlag; add(WFlagNumericWiden)
+    object WFlagOctalLiteral extends WFlag; add(WFlagOctalLiteral)
```
This is the one that caught my attention, as it is a very narrow warning. And well-meaning if misguided contributors add annoying warnings all the time that I probably want to suppress because they are so annoying.

lrytz replied:
This is the best I could come up with.. We cannot put `add(this)` it in the constructor of `WarningCategory`, as the objects here are evaluated lazily.

som-snytt commented on src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:
```diff
@@ -34,21 +34,31 @@ trait StandardScalaSettings {
+  val deprecation =    BooleanSetting ("-deprecation", "Emit warning and location for usages of deprecated APIs.") withAbbreviation "--deprecation" withPostSetHook { s =>
+    if (s.value) Wconf.tryToSet(List(s"cat=deprecation:w"))
+    else Wconf.tryToSet(List(s"cat=deprecation:s"))
+  }
```
There's `-Xlint:deprecation` which currently enables this setting. I guess `reflect.MutableSettings` is where setting `value` invokes the `postSetHook`.

som-snytt commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -24,6 +26,77 @@ trait Warnings {
+         |where <filters> is one or more <filter> combined with &, i.e., <filter>&...&<filter>
+         |
+         |Note: Run with `-Wconf:any:warning-verbose` to print warnings with their the category,
+         |site, origin (for deprecations) and since version (for deprecations).
```
typo `their category`, suggest `print warnings with their category and site, and for deprecations, their origin and since version.`  What's origin? oh I guess the library name thing in the message, or the actual class? OK I see below it's the latter. You can say `since="mylib 1.0"` but I don't know if that is common usage (or used at all).

lrytz replied:
> the library name thing in the message, or the actual class? OK I see below it's the latter

Yes, it's the full name of the deprecated symbol


> You can say `since="mylib 1.0"`

That is supported, in the sense that `mylib ` is dropped when parsing the `since` version. To filter deprecations for a specific library, use the `origin`, add a `since` filter if needed.

lrytz commented:
Talked a bit about this with @adriaanm the other day, one change we agreed to do is cut down the long list of warning categories (https://github.com/lrytz/scala/blob/confwarn/src/compiler/scala/tools/nsc/Reporting.scala#L310-L377) to something that can be supported by both scalac and dotc.

@SethTisue a review would still be highly appreciated :-)

ghik commented:
@lrytz maybe it would be worth to also add a line content filter? This would be useful for silencing unused imports. Relevan silencer issue: https://github.com/ghik/silencer/issues/42

lrytz commented:
That's a nice idea 👍

lrytz commented:
I thought a bit about reducing the number of warning categories. Currently we have plain categories like `deprecation` and `unchecked`, but there are also some with detailed sub-categories:

```
 - feature, feature-dynamics, feature-existentials, ...
 - lint, lint-adapted-args, lint-constant, lint-delayedinit-select, ...
 - unused, unused-imports, unused-locals, ...
 - w-flag, w-flag-dead-code, w-flag-extra-implicit, ...
```

For silencing, having these sub-categories is not really useful: if you don't want certain lint/unused/w-flag warnings, then don't enable them. For features, add the feature flag or import.

However, there's also the possibility of promoting warnings to errors (selectively, instead of using `-Werror`). For this case, having detailed categories would be very useful to ensure certain rules in a codebase, for example `-Wconf:cat=w-flag-numeric-widen:e,cat=unused-params:e,lint-infer-any:e,feature-dynamics:e`.

For the [previously mentioned motivation](https://github.com/scala/scala/pull/8373#issuecomment-546837185) of having fewer categories that can be supported by both Scala 2 and 3, a common subset of core categories should be fine too. It's not hard to customize compiler flags depending on the Scala version (cross-building projects are used to this).

A downside of adding many categories now is that it will require deprecation to remove them later if we ever want to.

Overall, my tendency currently is to leave the categories as they currently are in this PR.

SethTisue commented:
first: huge, huge thanks to @ghik for his pioneering work in this area

second: I pushed a merge with current HEAD, hopefully correctly (@lrytz you might want to review it briefly).

> Unchecked warnings are now all shown by default

I am hugely in favor of this 👏👏👏

okay, now let the nitpicking begin:

Naming: I have doubts that `@silent` is the best name. Taken out of context, it could mean many different things. I think it will occur rarely enough in most codebases that many readers encountering it will be unfamiliar with it. Including the word "warn" in the name seems desirable to me. Maybe `@nowarn`? (Java calls it `@SuppressWarnings`, but that's not our usual naming style.)

in `Site where the warning is triggered: site=my\\.package\\..* --The regex must match the full name of the warning position`, is the position always a package, or might it include e.g. a trailing class name?

The difference between "warning" and "info" isn't explained in the help text. I wasn't aware that this "info" category even existed in the compiler. If I'm not running with `-Werror`, is there any difference between an "info" and a "warning"? (I imagine I could get more insight into this by looking through the diffs, but I'm asking these questions not just for myself, but to suggest that the doc needs improving in this area.)

The categories seem fine to me. It's a hard thing to have definite opinions about. But it seems like the kind of thing that will be tweakable later in response to user feedback.

> The -deprecation, -feature and -unchecked settings are no longer directly used in the compiler, they are shortcuts for specific Wconf configurations. the compiler only looks at -Wconf

Should the documentation for these flags be updated to say what `-Wconf` invocations they are short for? That would help educate users about how this all works.

Thought: should one of the available actions be "summarize"? (Does that already exist internally in order to support the default behavior when `-deprecation` and/or `-feature` aren't specified?)

> Adds a -rootdir compiler flag

The doc for the flag should say what effect it has, should say why I would want to use it.

This is a bit tangential, but needing to import members of `scala.annotation`, or needing to write e.g. `@annotation.silent`, has always annoyed me. Could they be automatically imported? Why aren't they already? (Should I take this question elsewhere?)

I wonder a bit if users who cross-compile might want something in `@silent` that would let them specify which Scala versions it applies to. (I suppose we could wait and see if they ask.)

Okay, I just shook my brain a few more times and nothing further came out. Having this land is going to be AWESOME.

som-snytt commented:
Probably I'm the only one who wants to name the option `-Woof`.

SethTisue commented:
oh hi Som

som-snytt commented:
`-Woof`.

hrhino commented:
`-Whew`.

som-snytt commented:
`import annotation._` deserves its own issue. (`-Yimport:annotation` in current syntax.)

I haven't had a chance to exercise this in anger or jest, but I'll try to give it a shot.

I don't know how lrytz has time to try stuff, do a Spree, I bet he skis too.

For me, the hard part is the config. We just had this convo at the office, internal vs external config for warning suppression. I hate it all. I just want the build to tell me incrementally if my actions have incurred a problem. I don't care about the constant signal since 2.11.0.

If someone blesses the 2.12.0 build, then I don't care about those warnings at all, unless I go, Please tell me *all* the ****. Once you published 2.12.0 with those warnings, you basically said they didn't really matter. No more config is required.

lrytz commented:
> I bet he skis too

I used to, nowadays it's only being jealous of [people like this](https://twitter.com/jboner/status/1230148217184116739)

SethTisue commented:
/rebuild

SethTisue commented:
there was a performance issue reported against the silencer plugin today: https://github.com/ghik/silencer/issues/45

just now I tested this PR (2.13.2-bin-59cf05f-SNAPSHOT) with the test repo from that ticket and the same problem didn't occur :+1:

SethTisue commented:
> import annotation._ deserves its own issue

https://github.com/scala/bug/issues/11888

lrytz commented:
Note to self: silencer doesn't go into macro expansions anymore by default to search for `@silent` annotations (since https://github.com/ghik/silencer/commit/e10e610221e10f6e07eea3df7038f0cc7c173c8e / https://github.com/ghik/silencer/issues/45#issuecomment-589606709). In our case, `@silent` annotations are handled when they are type checked, and the corresponding tree is used to get the position range covered by the annotation. This should also work for `@silent` annotations within macro-expanded code (add a test?).

lihaoyi-databricks commented:
Bit of feedback on this PR:

- The annotation `@silent` is a bit too terse. Silence what? Logging? Exceptions? It's not clear to an uninformed reader, whereas if we called it `@silenceWarnings` that makes it abundantly obvious what's going on

- Java already provides a `java.lang.SuppressWarnings("hello")` annotation; can we leverage that? Would save an import, already perfectly named, and allows consistency across the platform

lrytz commented:
I agree that `@silent` is not a bit too generic, but I prefer @SethTisue's suggestion to use `@nowarn` over re-using `@SuppressWarnings` from the JDK, since it would have slightly different semantics when used in Scala vs Java.

Also, `@SuppressWarnings` takes an `Array[String]` parameter, which means you'd have to use it as `@SuppressWarnings(Array("filter"))` in Scala, that's quite cumbersome. Java seems to have a special case to allow `@SuppressWarnings("arg")` instead of ``@SuppressWarnings({"arg"})`, but we don't do that in Scala.

SethTisue commented:
I asked the Dotty team if they had feedback; in response, @smarter suggested calling attention to it on Discourse, which I have now done: https://contributors.scala-lang.org/t/configurable-warnings-warning-suppression/4096

lrytz committed:
minor cleanups

lrytz committed:
-Wconf compiler flag for configuring compiler warnings

Warnings are assigned a category.

Warnings can be filtered by category, by message regex, by site where
they are issued, and by source path. Deprecations can additionally be
filtered by origin (deprecated definition) and `since` version.

Filtered warnings can be reported as error, warning, info, summary
(like deprecations) or silent.

Adds a `-rootdir` compiler flag. It is used to relativize file paths
when using source filters (`-Wconf:src=some/source/File.scala:s`).
There might be other uses for it in the future.

Unchecked warnings are now all shown by default (they were summarized
like deprecations before).

The `-deprecation`, `-feature` and `-unchecked` settings are no longer
directly used in the compiler, they are shortcuts for specific `Wconf`
configurations. The compiler only looks at `-Wconf`.

Message filtering is performed in `global.currentRun.reporting`, not
in the `Global.reporter`. The reasons are:
  - Separation of concerns: `Reporter`s worry about how to do reporting,
    removing duplicate messages.
  - Custom `Reporter`s are used by sbt, silencer, REPL, etc. It's too
    hard to do the necessary changes to the `Reporter` interface.
  - The `Wconf` setting could change between compiler runs.
    `currentRun.reporting` respects those changes.

So all warnings in the compiler should go through `global.runReporting`
(which is the same as `global.currentRun.reporting`). This method takes
four parameters: pos, msg, category (new), site (new). The site is
usually `context.owner` (in the frontend) or `currentOwner` (in
transformations).

`Context` has a `warn` method with 3 parameters (pos, msg, category) and
inserts the `owner` as `site`, so this is used in the frontend (context
reporters are also used to support silent mode / trying twice).

The `global.warning` method is deprecated and no longer used.

There are a few calls to `Reporter.warning` left in the codebase where
no `runReporting` is reachable, I think they are all OK not to
categorize / allow filtering. E.g., when running Scaladoc

```
reporter.warning(null, "Plugins are not available when using Scaladoc")
```

lrytz committed:
Add `@nowarn` annotation for local warning suppression

Integrate the fantastic [silencer](https://github.com/ghik/silencer)
compiler plugin by @ghik into the compiler, which allows suppressing
warnings locally using he `@nowarn` annotation.

The `@nowarn` annotation suppresses warnings within the scope covered by
the annotation.
  - `@nowarn def foo = ...`, `@nowarn class C { ... }` suppress
    warnings in a definition
  - `expression: @nowarn` suppress warnings in a specific expression

The annotation can be configured to filter selected warnings, for
example `@nowarn("cat=deprecation")` only suppresses deprecation
warnings. The filter configuration syntax is the same as in `-Wconf`.

MiMa exception for addition of `scala.annotation.nowarn`

Reporting warnings is now delayed until after typer, because the typer
collects `@nowarn` annotations.

If we stop before typer (e.g., because there are errors in the parser),
all warnings are shown, even if they are enclosed in `@silent`.

If a phase before typer has both errors and warnings, all errors are
printed before the warnings.

lrytz committed:
gitignore for metals

lrytz commented:
> Maybe `@nowarn`?

done!

> is the position always a package, or might it include e.g. a trailing class name?

I'll improve the explanation. It's the full path to the enclosing definition, `p.C.f` in the example:
```
$> cat Test.scala
package p
class C {
  def f = 1 → 2
}
$> qsc Test.scala -Wconf:any:warning-verbose
Test.scala:3: warning: [deprecation @ p.C.f | origin=scala.Predef.ArrowAssoc.→ | version=2.13.0] method → in class ArrowAssoc is deprecated (since 2.13.0): Use `->` instead. If you still wish to display it as one character, consider using a font with programming ligatures such as Fira Code.
  def f = 1 → 2
            ^
1 warning
```

> difference between "warning" and "info" isn't explained in the help text

will do


> Should the documentation for these flags [deprecation/feature/unchecked] be updated

will do

> should one of the available actions be "summarize"?

that's already in there, we have `warning-summary` and `info-summary` actions

> doc for the flag [rootdir] should say what effect it has

will do

> specify which Scala versions it applies to

good idea. using `util.Properties.versionString`? can be a follow-up as you suggest.

lrytz committed:
fix some warning categories

lrytz committed:
better Wconf doc

lrytz committed:
empty filter is no filter

lrytz committed:
support -Wconf in repl

lrytz committed:
use full path in site and origin strings

SethTisue commented:
>> specify which Scala versions it applies to
> good idea. using util.Properties.versionString? can be a follow-up as you suggest.

Sorry, not sure what you're asking me for here.

lrytz commented:
I squashed all changes and split them up in 3 commits
  - small cleanups
  - `-Wconf`
  - silencer (`@nowarn`)

I also added a few small cleanups and improvements after that, based on the review feedback. I think we can merge this now :-)

Follow-up changes that were suggested:
  - Support filtering based on the source code text at the line where the message is issued (useful for unused imports, for example)
  - Support filtering based on the scala version

ijuma commented:
Awesome!

ijuma commented:
I tested this feature in Apache Kafka and I have a branch with 0 warnings and `-Xfatal-warnings` enabled. This is awesome. :) Two observations:

1. It would be really useful for the `nowarn` annotation to be added to Scala 2.12 even if it doesn't do anything. Kafka, like most projects, supports Scala 2.12 and Scala 2.13. I am fine with compiling with Scala 2.13 by default and having `-Xfatal-warnings` enabled for that version only, but I can't easily use the annotation if it's not present at all in Scala 2.12.

2. I was unable to get the `site` filter to work via a compilation argument. I had to fallback to the `source` filter instead, which is less precise. This could be user error, but I noticed that the tests seemingly only cover the default package:

> check(infos(code, "site=A.f:i"), Nil)
> check(infos(code, "site=A.invokeDeprecated:i"), List(l5a, l5b))

I will try this on a minimized test case and report an issue (hopefully soon), but I wanted to give a heads up since 2.13.2 has not been released yet.

lrytz commented:
> `nowarn` annotation to be added to Scala 2.12

That's certainly an option. Alternatively, adding it to https://github.com/scala/scala-collection-compat would be lower overhead - would it be an option to add this dependency?

For the site filter, it seems to work for me; is it maybe that the leftmost matching rule defines the action?

```
$> cat Test.scala
package a.pu
class K {
  def f = 1 → 2
  @annotation.nowarn("site=a.pu.K.g")
  def g = 1 → 2
}
$>

$> scv 2.13.2-bin-cd56924 -Wconf:cat=deprecation:wv Test.scala
Test.scala:4: warning: [deprecation @ a.pu.K.f | origin=scala.Predef.ArrowAssoc.→ | version=2.13.0] method → in class ArrowAssoc is deprecated (since 2.13.0): Use `->` instead. If you still wish to display it as one character, consider using a font with programming ligatures such as Fira Code.
  def f = 1 → 2
            ^
1 warning
$>

$> scv 2.13.2-bin-cd56924 -Wconf:site=a.pu.K.f:s,cat=deprecation:wv Test.scala
$>

$> scv 2.13.2-bin-cd56924 -Wconf:cat=deprecation:wv,site=a.pu.K.f:s Test.scala
Test.scala:4: warning: [deprecation @ a.pu.K.f | origin=scala.Predef.ArrowAssoc.→ | version=2.13.0] method → in class ArrowAssoc is deprecated (since 2.13.0): Use `->` instead. If you still wish to display it as one character, consider using a font with programming ligatures such as Fira Code.
  def f = 1 → 2
            ^
1 warning
$>
```

ijuma commented:
@lrytz Thanks for the quick response.

`scala-collection-compat` would be perfect, we already use it for other reasons.

I got `site` to work with the help from your example. I got confused because the help text in the PR description shows a `\` before the `.`:

![image](https://user-images.githubusercontent.com/24747/78258951-daac1c00-74b0-11ea-9d89-d2a0b1476066.png)

lrytz commented:
OK, I see how the `\` can be annoying to get right... The thing is, the pattern is a java regex, so if you write `site=my.pack.*`, it matches `myApack`.

The following command line worked as expected for me:

```
scv 2.13.2-bin-cd56924 '-Wconf:site=a\.pu\.K\.f:s,cat=deprecation:wv' Test.scala
```

ijuma commented:
@lrytz I investigated more and I found the issue. I got a bit unlucky and the first warning I tried to suppress didn't seem to have the `@` location captured correctly:

> "Error:(109, 68) [unused-params @ <root>] parameter value nn in method body is never used
>     def body[T <: AbstractRequest](implicit classTag: ClassTag[T], nn: NotNothing[T]): T = {"

In the Kafka codebase, we had one other example like the above (similar pattern), but all other ones worked fine. The usage of `wv` from your example helped track this down.

lrytz commented:
I'll take a look if I can reproduce and fix that.

For `@nowarn`, I submitted https://github.com/scala/scala-collection-compat/pull/312

lrytz commented:
In the meantime, until the annotation finally ends up in 2.12.12, we could suggest cross-building projects to use version-specific source files and
  - define a type alias in 2.13
  - define their own `nowarn` annotation class in 2.11/2.12


The 2.13 version would be

```scala
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

package my { package object project { type nowarn = scala.annotation.nowarn } }

// Exiting paste mode, now interpreting.


scala> @my.project.nowarn def foo = 1 → 2
def foo: (Int, Int)
```

I think there's also a way to make it work with a wildcard import, maybe even without version-specific source files? @sjrd does these kind of tricks I believe.

lrytz commented:
@ijuma https://github.com/scala/scala/pull/8852 fixes the warning site for unused warnings.

ijuma commented:
Thanks!

charpov commented:
``comparing a fresh object using `eq` will always yield false`` is being suppressed by `@nowarn("cat=other")` but not by any of the `other-` subcategories I have tried.  Is there one that is missing from the documentation?

som-snytt commented:
@charpov https://github.com/scala/scala/blob/v2.13.3/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala#L1046 just says other. (Probably I'd ask on gitter instead of the old PR.)
